### PR TITLE
feat(levm): ethereum rust integration

### PR DESCRIPTION
**Pull request by JereSalo**
**Motivation**

Eventually, we will stop using revm and replace it with levm.

This PR aims to be the first step. It aims to prepare the current codebase so that both implementations can coexist while levm continues to be developed and does not disturb the rest of the development. When the time to remove revm comes, it'd be easier with this proposal since it'd require removing the code under the #[cfg(not(feature = "levm"))] and after that, removing the feature flag definitively.

**Description**

- Adds the `levm` feature flag to the `vm` and `blockchain` packages, and to the `ethereum_rust` binary.
- In `vm`, add `execute_block` implementation that uses `levm` and use it when `levm` feature flag is set. Use the original implementation if not.
- In `blockchain`, add `add_block` implementation that uses `levm` and use it when `levm` feature flag is set. Use the original implementation if not.

### CODE REVIEW

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (fkrause98) [inline crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:162]: If this is still pending, open an issue, and reference it here.

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (fkrause98) [inline crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:183]: Do not unwrap, handle this.

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (fkrause98) [inline crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:185]: Why? This seems arbitrary, let's name these constants and comment where they come from.

File: crates/vm/levm/src/opcode_handlers/system.rs
Comment (fkrause98) [inline crates/vm/levm/src/opcode_handlers/system.rs:97]: Create the issue for this, and link it here or add the issue's number in this comment

File: crates/vm/levm/src/opcode_handlers/system.rs
Comment (fkrause98) [inline crates/vm/levm/src/opcode_handlers/system.rs:105]: Maybe out of the scope of this PR, but maybe we can also handle these unwraps below.

File: crates/vm/levm/src/opcode_handlers/system.rs
Comment (fkrause98) [inline crates/vm/levm/src/opcode_handlers/system.rs:161]: Open an issue for this and link it.

Comment (fkrause98) [reply crates/vm/levm/src/opcode_handlers/system.rs:161]: The same goes for each todo below.

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:104]: This function returns result, handle this unwrap properly.

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:111]: Ditto

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:117]: Open an issue for this

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:163]: Do we need this debug statement?

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:168]: Nit: These transactions seem to overlap, consider using a '|' for a 'or' match, like in this [example](https://doc.rust-lang.org/rust-by-example/flow_control/match.html).

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:170]: Can we handle these unwraps? This function already returns Result

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:194]: Ditto

File: crates/vm/vm.rs
Comment (fkrause98) [inline crates/vm/vm.rs:180]: Keep in mind: This operations use u64s, what happens if there is an overflow? Can it happen? What should we do in such case? Depending on the behavior we want, we can use a dedicated operation for this like saturating_add, wrapping_add, etc..

File: crates/common/types/transaction.rs
Comment (JereSalo) [reply crates/common/types/transaction.rs:1597]: Because when running the L2 we had this error with the alias on:
`Error: server returned an error response: error code -32000: Invalid params: duplicate field "input"`
So removing it fixed that for us. I still don't know why it happens though.

File: crates/vm/db.rs
Comment (ilitteri) [reply crates/vm/db.rs:42]: Unwraps are spread all over `levm's` codebase. As is involves a lot of work (unrelated to this PR), we've filed an issue that will be tackled immediately after this one is merged.

File: crates/vm/db.rs
Comment (ilitteri) [reply crates/vm/db.rs:47]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920

File: crates/vm/levm/src/opcode_handlers/environment.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/environment.rs:118]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:162]: 🫡

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:183]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:185]: Agreed.

File: crates/vm/levm/src/opcode_handlers/system.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/system.rs:97]: 🫡

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs:104]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs:111]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs:170]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs:180]: Agreed, tracked this in https://github.com/lambdaclass/lambda_ethereum_rust/issues/1075, it'll be done after removing the `unwrap`s and `expect`s so we have a solid error handling basis.

**Review by fkrause98 — APPROVED** [review]
Looks ok, unwraps and error handling will be handled in a following PR/issue, approving so we can continue.


One thing though: Let's make sure someone else also reviews this PR.

File: crates/blockchain/blockchain.rs
Comment (ilitteri) [reply crates/blockchain/blockchain.rs]: All the `levm` `dbg!` macros where commented out (they're still being used), and the ones form `ethereum_rust` were removed [8c21a37](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/8c21a37879b1704836faca49fb0a6828b1ce174f).

File: crates/blockchain/blockchain.rs
Comment (ilitteri) [reply crates/blockchain/blockchain.rs]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029#discussion_r1831061421.

File: crates/blockchain/payload.rs
Comment (ilitteri) [reply crates/blockchain/payload.rs]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029#discussion_r1831061421

File: crates/vm/db.rs
Comment (ilitteri) [reply crates/vm/db.rs:30]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029#discussion_r1829787920.

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs]: Done in [1ff2d77](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/1ff2d77272f33be8879304c3e849b0b39c95df9d).

File: crates/vm/levm/src/opcode_handlers/system.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/system.rs]: Done in [1ff2d77](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/1ff2d77272f33be8879304c3e849b0b39c95df9d).

File: crates/vm/levm/src/opcode_handlers/system.rs
Comment (ilitteri) [reply crates/vm/levm/src/opcode_handlers/system.rs]: Done in [1ff2d77](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/1ff2d77272f33be8879304c3e849b0b39c95df9d).

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs]: Done in [1ff2d77](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/1ff2d77272f33be8879304c3e849b0b39c95df9d).

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs]: Nope, it was removed.

File: crates/vm/vm.rs
Comment (ilitteri) [reply crates/vm/vm.rs:192]: Same as https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/files#r1829787920.

**Comment by ilitteri** [comment]
> Looks ok, unwraps and error handling will be handled in a following PR/issue, approving so we can continue.
> 
> One thing though: Let's make sure someone else also reviews this PR.

Thanks for the review! As a TLDR for the comments, I've responded:
- `dbg!` macros were either commented out or removed (the `levm` ones were commented out because they're still being used in dev) [8c21a37](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/8c21a37879b1704836faca49fb0a6828b1ce174f).
- Issues for the TODOs were filed and linked in the comments [1ff2d77](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1029/commits/1ff2d77272f33be8879304c3e849b0b39c95df9d).
- An issue to handle the unwraps, expects and more panickable code was filed https://github.com/lambdaclass/lambda_ethereum_rust/issues/1074.
- An issue to handle overflows in arithmetic operations was filed https://github.com/lambdaclass/lambda_ethereum_rust/issues/1075.

File: crates/common/types/transaction.rs
Comment (rodrigo-o) [reply crates/common/types/transaction.rs:1597]: This was added in this PR: https://github.com/lambdaclass/lambda_ethereum_rust/pull/915 to fix an issue with EIP1559 transactions that required data instead of input apparently. Unfortunately we don't have a test to check this regression. Maybe the error happened on a json with both input and data?

**Review by rodrigo-o — APPROVED** [review]
LGTM given that we already have issues for Fran's comments. I left just a small comment regarding the alias removal.

Apart from that, I try to avoid leaving lines commented  (like the `dbg!` ones) but if others are okay with it I am.

**Pull request merged**
